### ISSUE TITLE
Detection not same as with python with same model and picture

### ISSUE THREAD

**Issue by FalcoGer** [migueldeicaza/TensorFlowSharp#191]
When using the python object detection script that comes with the object detection api. (slightly altered to fetch different images and my own trained model)
and when using the example program for object detection with the same input images and the same model and the same label map, then there is a major discrepency.
I can give you a json I am creating with the python script and some debug info I create with the c# program for the same image:

python:
```
[
  {
    "image_path": "./100percent.jpg"
  },
  {
    "boundingBoxes": [
      [
        {
          "xmin": 229,
          "ymin": 104,
          "xmax": 320,
          "ymax": 529
        },
        {
          "probability": 0.9999961853027344,
          "detection_class": 1,
          "label": "Feature 1"
        }
      ],
      [
        {
          "xmin": 409,
          "ymin": 134,
          "xmax": 448,
          "ymax": 330
        },
        {
          "probability": 0.9997336268424988,
          "detection_class": 2,
          "label": "Feature 2"
        }
      ],
      [
        {
          "xmin": 340,
          "ymin": 96,
          "xmax": 481,
          "ymax": 647
        },
        {
          "probability": 0.9950695037841797,
          "detection_class": 3,
          "label": "Feature 3"
        }
      ]
    ]
  },
  {
    "shape": {
      "dimx": 820,
      "dimy": 700
    }
  }
]
```

Here is the result from TensorflowSharp. Only one feature was detected with a different confidence.
```
Processing picture: ./100percent.jpg
Class: Feature 1[1]
xmin: 0,147357          ymin: 0,2830129         xmax: 0,7979434         ymax: 0,402808
Confidence: 95,5%
```

The overall detection rate is about 10% of that using python, with most pictures getting no detection at all, and some pictures getting just one out of three features. Also the bounding boxes vary greatly between what is detected by both python and c#.
I suspect the decoding of the jpg returns different arrays, but this needs more testing.
For python I use cv2 to load the image into a numpy array, but I get the same results with tensorflow's decode jpg function. for c# i use basically use what is given in the example.

Unfortunately I do not have time to do much more than to write an issue here and have to make a workaround with python for now.

As a note, this can also be seen in the test image. only 2 kites are detected, while python would detect nearly all of them, along with the people.

**Comment by migueldeicaza** [comment]
This is a straight binding of the lower-level engine of TensorFlow, as you point out there could be a number of differences that affect the results.

Your best bet is to provide a self-contained test case for someone to investigate.

**Comment by IlSaCCo** [comment]
Hey Miguel, 

I am having the same kind of problem FalcoGer is having: I trained my custom model with tensorflow, froze it, tested it using the object_detection_tutorial.ipynb that ship with the tensorflow api (adjusting the model name and paths) and I get an 90% accuracy on 10 test images.

Using the objectdetection# sample project, the same protobuf model, on the same 10 jpg I get only 1 object in 1 image detected properly. I debugged the sample code line by line and I do not see any problem but when it comes to unmanaged array there is not much I can see apart a pointer.

**Comment by Robertb84** [comment]
Same as written in [`#205`](https://github.com/migueldeicaza/TensorFlowSharp/issues/205) . 

MobileNet work better but well my next problem is that i can't port it to mobile (because of DecodeJpeg)

**Comment by matiasmolinas** [comment]
The same problem here, with a custom model I have different results from python and from C#, from C# I can't detect objects that the same model is detecting from python.

**Comment by epignatelli** [comment]
I am having the same problem with a custom network as well.
Classifications are more or less random when using the C# code.

To add a variable, my model was stored in a `hdf5` file, due to keras.
Could the conversion process to a tensorflow `pb` format have broken the model?

What was your workflow?

**Comment by johnlundberg** [comment]
I am also having this issue.

**Comment by sallerga** [comment]
I have noticed that the DecodeJpeg function does not produce the same results for c# and python
If I run the following code on the attached file, the resulting images are not identical.
Any idea why? Can it cause the different inference results?
```csharp
using System.Text;
using TensorFlow;
internal static void SaveCropedAndResizedFile(string source, string destination)
{
    using (var graph = new TFGraph())
    {
        var input = graph.Placeholder(TFDataType.String);

        var file = graph.ReadFile(input);
        var decodedeJpg = graph.DecodeJpeg(contents: file, channels: 3,
            dct_method: "INTEGER_ACCURATE");

        var png = graph.EncodePng(decodedeJpg, compression: 0);
        var writeFile = graph.WriteFile(
            graph.Const(TFTensor.CreateString(
                Encoding.ASCII.GetBytes(destination)), TFDataType.String),
            png);

        var fileNameTensor = TFTensor.CreateString(
            Encoding.ASCII.GetBytes(source));
        using (var session = new TFSession(graph))
        {
            session.Run(
                inputs: new[] { input },
                inputValues: new[] { fileNameTensor },
                outputs: new[] { png },
                targetOpers: new[] { writeFile }
                );
        }
    }
}
```
```python
import tensorflow as tf

def decode_file(source, destination):    
    with tf.Graph().as_default():
        input = tf.placeholder(tf.string)
        contents = tf.read_file(input)
        jpg_image = tf.image.decode_jpeg(contents, channels=3, dct_method= "INTEGER_ACCURATE")

        encoded_png = tf.image.encode_png(jpg_image,0)
        
        write_file = tf.write_file(tf.constant(destination), encoded_png)

        with tf.Session() as sess:
            sess.run(write_file, feed_dict={input:source})

```
![testimage](https://user-images.githubusercontent.com/5845709/39124840-391543e0-46fd-11e8-99f3-8569dbe8babb.jpg)

**Comment by ChrisQlasty** [comment]
The same here: when I put random input to my .pb in python it can change the output (predicted class number) within script runs - but in C# no matter what the input is (random or true data) it always returns 1.

Does anyone have updates on this thread?

Python code:

```python
import tensorflow as tf
import numpy as np
from tensorflow.python.platform import gfile

with tf.Graph().as_default() as graph: # Set default graph as graph
    with tf.Session() as sess:
        with gfile.FastGFile("frozen_model.pb", 'rb') as f:

            graph_def = tf.GraphDef()
            graph_def.ParseFromString(f.read())
            sess.graph.as_default()

            tf.import_graph_def(graph_def, input_map=None, return_elements=None, name="", op_dict=None, producer_op_list=None)
            
            X_in = graph.get_tensor_by_name('X:0') # Input Tensor
            l_in = graph.get_tensor_by_name('seq_length:0')  # Input Tensor
            my_op = graph.get_tensor_by_name('top_result:0') # Output Tensor
            seq_len = 25

            for k in range(5):
                inp =  np.random.random((1,166,8))*2-1
                print("min: {0:.3} max: {1:.3} mean: {2:.3}".format( np.min(inp), np.max(inp), np.mean(inp)))

                Session_out = sess.run(my_op, feed_dict = {X_in : inp, l_in:[seq_len]})
                print(Session_out[0])
```
The output is:
```text
min: -0.999 max: 0.998 mean: 0.00342  
26  
min: -0.999 max: 1.0 mean: 0.00399  
23  
min: -0.999 max: 0.999 mean: -0.00744  
17
min: -0.993 max: 0.997 mean: 0.0385
17
min: -0.998 max: 0.999 mean: -0.00515
17
```

C# method code:

```c#
Random r = new Random();
int seq_len = 25;

for (int k = 0; k < 5; k++)
{                
	float[,,] testData = new float[1, 166, 8];
	for (int i = 0; i < 8; i++)	
		for (int j = 0; j < 166; j++)		
			testData[0, j, i] = (float)r.NextDouble() * 2.0f - 1.0f;
			
	Console.WriteLine(String.Format("min: {0} max: {1} mean: {2}", testData.Cast<float>().Min(), testData.Cast<float>().Max(), testData.Cast<float>().Sum() / testData.Cast<float>().Count()));

	string ans = "";
	using (var graph = new TFGraph())
	{
		graph.Import(File.ReadAllBytes("frozen_model.pb"),"");
		var session = new TFSession(graph);
		var runner = session.GetRunner();                    
		
		runner.AddInput(graph["X"][0], new TFTensor(testData));
		runner.AddInput(graph["seq_length"][0], new TFTensor(seq_len));
		runner.Fetch(graph["top_result"][0]);		
		var output = runner.Run();
		
		TFTensor result = output[0];
		ans = result.ToString();
		Console.WriteLine(ans);
	}                
}
```
Output:
```text
min: -0,9999681 max: 0,999267 mean: 0,006615811
[1]
min: -0,9987128 max: 0,9970865 mean: -0,002892239
[1]
min: -0,9991289 max: 0,9989161 mean: 0,002840083
[1]
min: -0,9931433 max: 0,9993781 mean: -0,00225115
[1]
min: -0,9966126 max: 0,9995595 mean: -0,01307642
[1]
```

**Comment by epignatelli** [comment]
I haven't tested further, but I can confirm that the problem is in the JPEG decompression to tensor.
Experiencing this even just within python environment, two different numpy versions yield different results.

**Comment by edward-** [comment]
I have the same problem, Help me!, please.

I suspect it is UtilImage but I can not solve it.

**Comment by Amanpradhan** [comment]
Any one of you have solved this problem as I am suffering from the same?

**Comment by axeldietz** [comment]
Same problem here.

In order to exclude the JPG decompression influence, I used a PNG lossless compression file format test image.

How cum this problem has not been resolved? It is general known: 

From: https://www.died.tw/2017/12/tensorflow-object-detection-api-test-in.html:

Using same model: "ssd_mobilenet_v1_coco_2017_11_17", in TensorflowSharp you can't get same result as official tensorflow, so if you want to develop in TensorflowSharp, be careful with it!!

This way it makes no sense to use TensorFlowSharp, which is a pity.

**Comment by mg021** [comment]
@FalcoGer 
I Have two Result 
with python My Model  Work Well .
with TensorflowSharp  my model Is not Goood .
Help Me .
Is TensorflowSharp Ready or ...??
